### PR TITLE
ddl: improve backfill more general

### LINKED ISSUE #41002: Change the table structure to be more general

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #41002

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/constant.go [hunk 1]
```diff
@@ -38,8 +38,8 @@ const (
 	HistoryTableID = meta.MaxInt48 - 3
 	// MDLTableID is the table ID of `tidb_mdl_info`.
 	MDLTableID = meta.MaxInt48 - 4
-	// BackfillTableID is the table ID of `tidb_ddl_backfill`.
-	BackfillTableID = meta.MaxInt48 - 5
+	// BackgroundSubtasksTableID is the table ID of `tidb_background_subtasks`.
+	BackgroundSubtasksTableID = meta.MaxInt48 - 5
 	// BackfillHistoryTableID is the table ID of `tidb_ddl_backfill_history`.
 	BackfillHistoryTableID = meta.MaxInt48 - 6
```
Comment (Benjamin2037) [inline ddl/constant.go:44]: This 
```suggestion
	BackgroundSubTaskHistory = meta.MaxInt48 - 6
```

Comment (Benjamin2037) [reply ddl/constant.go:44]: Please also change this history table according

Comment (hawkingrei) [reply ddl/constant.go]: Done

File: ddl/constant.go [hunk 2]
```diff
@@ -38,38 +38,34 @@ const (
 	HistoryTableID = meta.MaxInt48 - 3
 	// MDLTableID is the table ID of `tidb_mdl_info`.
 	MDLTableID = meta.MaxInt48 - 4
-	// BackfillTableID is the table ID of `tidb_ddl_backfill`.
-	BackfillTableID = meta.MaxInt48 - 5
-	// BackfillHistoryTableID is the table ID of `tidb_ddl_backfill_history`.
-	BackfillHistoryTableID = meta.MaxInt48 - 6
+	// BackgroundSubtasksTableID is the table ID of `tidb_background_subtasks`.
+	BackgroundSubtasksTableID = meta.MaxInt48 - 5
+	// BackgroundSubtaskHistoryTableID is the table ID of `tidb_background_subtask_history`.
```
Comment (Benjamin2037) [inline ddl/constant.go:43]:
```suggestion
	// BackgroundSubtasksHistoryTableID is the table ID of `tidb_background_subtasks_history`.
```

File: ddl/constant.go [hunk 3]
```diff
@@ -38,38 +38,34 @@ const (
...
+	BackgroundSubtaskHistoryTableID = meta.MaxInt48 - 6
```
Comment (Benjamin2037) [inline ddl/constant.go:44]:
```suggestion
	BackgroundSubtasksHistoryTableID = meta.MaxInt48 - 6
```

Comment (hawkingrei) [reply ddl/constant.go]: %s/BackgroundSubtaskHistoryTableID/BackgroundSubtasksHistoryTableID/g

add 's'.

File: ddl/constant.go (hunk 2 above)
Comment (hawkingrei) [reply ddl/constant.go]: BackgroundSubtaskHistory -> BackgroundSubtasksHistory

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: ddl/job_table_test.go [hunk 4]
```diff
@@ -292,6 +300,7 @@ func TestSimpleExecBackfillJobs(t *testing.T) {
 	// 1      jobID2     eleID3    ""
 	require.NoError(t, err)
 	// test get some backfill jobs
+	log.Info("debug", zap.Int64("jobid1", jobID2-1), zap.Int64("jobid2", jobID2+1))
```
Comment (zimulala) [inline ddl/job_table_test.go:303]: Do we need to remove this debug log?

File: ddl/constant.go [hunk 5]
```diff
@@ -38,38 +38,34 @@ const (
...
-		start_key blob,
-		end_key blob,
-		start_ts bigint,
-		finish_ts bigint,
-		row_count bigint,
-		backfill_meta longblob,
-		unique key(ddl_job_id, ele_id, ele_key(20), id))`
-	// BackfillHistoryTableSQL is the CREATE TABLE SQL of `tidb_ddl_backfill_history`.
-	BackfillHistoryTableSQL = "create table " + BackfillHistoryTable + `(
+		exec_id varchar(256) not null,
```
Comment (zimulala) [inline ddl/constant.go:59]: I think this can be nil, the first insert, it's is nil.

File: ddl/constant.go [hunk 6]
```diff
@@ -38,38 +38,34 @@ const (
...
+		exec_expired timestamp not null,
```
Comment (zimulala) [inline ddl/constant.go:60]: Ditto.

File: ddl/backfilling.go [hunk 7]
```diff
@@ -99,14 +99,11 @@ type BackfillJob struct {
...
 	// range info
-	CurrKey  []byte
-	StartKey []byte
-	EndKey   []byte
-
-	StartTS  uint64
-	FinishTS uint64
-	RowCount int64
-	Meta     *model.BackfillMeta
+	CurrKey       []byte
```
Comment (zimulala) [inline ddl/backfilling.go:102]: Could we put `CurrKey` to `Meta`?

File: ddl/backfilling.go [hunk 8]
```diff
@@ -99,14 +99,11 @@ type BackfillJob struct {
 	InstanceID    string
 	InstanceLease types.Time
 	// range info
```
Comment (zimulala) [inline ddl/backfilling.go:101]: Maybe we can remove this line.

File: ddl/job_table_test.go (hunk 4 above)
Comment (hawkingrei) [reply ddl/job_table_test.go]: Done

File: ddl/constant.go (hunk 5 above)
Comment (hawkingrei) [reply ddl/constant.go]: Done

File: ddl/constant.go (hunk 6 above)
Comment (hawkingrei) [reply ddl/constant.go]: Done

File: ddl/backfilling.go (hunk 8 above)
Comment (hawkingrei) [reply ddl/backfilling.go]: Done

File: ddl/constant.go [hunk 9]
```diff
@@ -38,38 +38,34 @@ const (
...
+		state varchar(64) not null,
+		checkpoint longblob not null,
+		start_time bigint not null,
+		state_update_time bigint not null,
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply ddl/constant.go]: Done

File: parser/model/ddl.go [hunk 10]
```diff
@@ -447,8 +447,11 @@ type BackfillMeta struct {
 	WarningsCount map[errors.ErrorID]int64         `json:"warnings_count"`
 	Location      *TimeZoneLocation                `json:"location"`
 	ReorgTp       ReorgType                        `json:"reorg_tp"`
-
-	*JobMeta `json:"job_meta"`
+	RowCount      int64                            `json:"row_count"`
+	StartKey      []byte                           `json:"start_key"`
+	EndKey        []byte                           `json:"end_key"`
+	CurrKey       []byte                           `json:"curr_key"`
```
Comment (zimulala) [inline parser/model/ddl.go:453]: Change `[]byte` to `kv.Key`? Unify it with the other key types.

File: ddl/constant.go [hunk 11]
```diff
@@ -25,10 +25,10 @@ const (
 	ReorgTable = "tidb_ddl_reorg"
 	// HistoryTable stores the history DDL jobs.
 	HistoryTable = "tidb_ddl_history"
-	// BackfillTable stores the information of backfill jobs.
-	BackfillTable = "tidb_ddl_backfill"
-	// BackfillHistoryTable stores the information of history backfill jobs.
-	BackfillHistoryTable = "tidb_ddl_backfill_history"
+	// BackgroundSubtasksTable stores the information of background jobs.
+	BackgroundSubtasksTable = "tidb_background_subtasks"
```
Comment (zimulala) [inline ddl/constant.go:29]:
```suggestion
	BackgroundSubtaskTable = "tidb_background_subtask"
```

We needn't `s`.
e.g.  `mysql.user` not `mysql.users`

File: ddl/constant.go [hunk 12]
```diff
@@ -25,10 +25,10 @@ const (
...
+	// BackgroundSubtasksHistoryTable stores the information of history backfill jobs.
+	BackgroundSubtasksHistoryTable = "tidb_background_subtasks_history"
```
Comment (zimulala) [inline ddl/constant.go:31]:
```suggestion
	BackgroundSubtaskHistoryTable = "tidb_background_subtask_history"
```

File: parser/model/ddl.go (hunk 10 above)
Comment (hawkingrei) [reply parser/model/ddl.go:453]: It is for ```parser``` to import tidb's package. it will happen cycle import.

File: ddl/job_table.go [hunk 13]
```diff
@@ -630,15 +631,36 @@ func getJobsBySQL(sess *session, tbl, condition string) ([]*model.Job, error) {
...
+		mateByte, err := bj.Meta.Encode()
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+
+		if i != 0 {
+			sqlBuilder.WriteString(", ")
+		}
+		sqlBuilder.WriteString(fmt.Sprintf("('%d_%d_%s_%d', %d, %d, '%s', '%s', '%s', %s, %d, %d, %s)",
+			bj.ID, bj.EleID, hex.EncodeToString(bj.EleKey), bj.JobID, bj.StoreID, bj.Tp, bj.InstanceID, bj.InstanceLease, bj.State.String(), wrapKey2String(bj.Meta.CurrKey),
```
Comment (zimulala) [inline ddl/job_table.go:654]: s/`bj.ID, bj.EleID, hex.EncodeToString(bj.EleKey), bj.JobID`/`bj.JobID,hex.EncodeToString(bj.EleKey), bj.EleID, bj.ID`?
We're going to use a prefix index?

File: ddl/constant.go [hunk 14]
```diff
@@ -38,38 +38,34 @@ const (
...
+		exec_id varchar(256),
+		exec_expired timestamp,
+		state varchar(64) not null,
+		checkpoint longblob not null,
+		start_time timestamp not null,
+		state_update_time timestamp not null,
+		meta longblob,
+		unique key(namespace, task_key))`
+	// BackgroundSubtaskHistoryTableSQL is the CREATE TABLE SQL of `tidb_background_subtask_history`.
+	BackgroundSubtaskHistoryTableSQL = "create table " + BackgroundSubtaskHistoryTable + `(
```
Comment (zimulala) [inline ddl/constant.go:68]: Could we put this table too?

**Review by zimulala — APPROVED** [review]

File: ddl/dist_owner.go [hunk 15]
```diff
@@ -222,26 +224,48 @@ func checkBackfillJobCount(sess *session, ddlJobID, currEleID int64, currEleKey
 	if err != nil {
 		return 0, errors.Trace(err)
 	}
-
-	backfillJobCnt, err = GetBackfillJobCount(sess, BackfillTable, fmt.Sprintf("ddl_job_id = %d and ele_id = %d and ele_key = %s",
-		ddlJobID, currEleID, wrapKey2String(currEleKey)), "check_backfill_job_count")
+	backfillJobCnt, err = GetBackfillJobCount(sess, BackgroundSubtaskTableSQL, fmt.Sprintf("task_key like `%d_%s_%d%%`",
+		ddlJobID, hex.EncodeToString(currEleKey), currEleID), "check_backfill_job_count")
```
Comment (zimulala) [inline ddl/dist_owner.go:228]:
```suggestion
	backfillJobCnt, err = GetBackfillJobCount(sess, BackgroundSubtaskTableSQL, fmt.Sprintf("task_key like `%%_%s_%d_%d`",
		hex.EncodeToString(currEleKey), currEleID, ddlJobID, descStr), "check_backfill_job_count")
```

Comment (zimulala) [reply ddl/dist_owner.go:228]: Then it is best to wrap into a generating function, only not easy to write wrong

Comment (hawkingrei) [reply ddl/dist_owner.go]: Done

File: ddl/index.go [hunk 16]
```diff
@@ -1367,8 +1367,8 @@ func (w *baseIndexWorker) UpdateTask(bfJob *BackfillJob) error {
 	s := newSession(w.backfillCtx.sessCtx)
 
 	return s.runInTxn(func(se *session) error {
-		jobs, err := GetBackfillJobs(se, BackfillTable, fmt.Sprintf("ddl_job_id = %d and ele_id = %d and ele_key = %s and id = %d",
-			bfJob.JobID, bfJob.EleID, wrapKey2String(bfJob.EleKey), bfJob.ID), "update_backfill_task")
+		jobs, err := GetBackfillJobs(se, BackgroundSubtaskTable, fmt.Sprintf("task_key = `%d_%s_%d_%d`",
```
Comment (zimulala) [inline ddl/index.go:1370]:
```suggestion
		jobs, err := GetBackfillJobs(se, BackgroundSubtaskTable, fmt.Sprintf("task_key = '%d_%s_%d_%d",
```

Comment (hawkingrei) [reply ddl/index.go]: Done

File: ddl/job_table.go [hunk 17]
```diff
@@ -632,8 +633,8 @@ func getJobsBySQL(sess *session, tbl, condition string) ([]*model.Job, error) {
 }
 
 func syncBackfillHistoryJobs(sess *session, uuid string, backfillJob *BackfillJob) error {
-	sql := fmt.Sprintf("update mysql.%s set state = %d where ddl_job_id = %d and ele_id = %d and ele_key = %s and exec_id = '%s' limit 1;",
-		BackfillHistoryTable, model.JobStateSynced, backfillJob.JobID, backfillJob.EleID, wrapKey2String(backfillJob.EleKey), uuid)
+	sql := fmt.Sprintf("update mysql.%s set state = %s where task_key like \"%d_%s_%d%%\" and exec_id = '%s' limit 1;",
```
Comment (zimulala) [inline ddl/job_table.go:636]:
```suggestion
	sql := fmt.Sprintf("update mysql.%s set state = %s where task_key like \"%d_%s_%d_%%\" and exec_id = '%s' limit 1;",
```

Comment (hawkingrei) [reply ddl/job_table.go]: Done

**Pull request merged**